src/tiledb/cloud/files/utils.py

```diff
@@ -264,3 +270,170 @@ def _auth_headers(cfg: configuration.Configuration) -> Mapping[str, object]:
+        report. Differently it returns an UploadFoldersResults attrs class.
+        Defaults to True.
+    :param logger: A logging.Logger instance, defaults to None.
+    :param verbose: Verbose logging, defaults to None
+    :return: A dictionary containing a report message
+        and an upload errors dictionary (if any)
+    """
+    logger = logger or get_logger_wrapper(verbose)
+    logger.info("=====")
+
+    if flatten:
+        raise NotImplementedError(
+            "The option to flatten a folder structure is not yet implemented"
+        )
+
+    # Prepare and sanitize arguments
+    output_uri = output_uri.strip("/")
+
+    input_uri = input_uri if input_uri.endswith(os.sep) else input_uri + os.sep
+    base_dir = os.path.dirname(input_uri)
+    base_dir = os.path.basename(base_dir)
+
+    namespace, name = utils.split_uri(output_uri)
+    _, sp, acn = groups._default_ns_path_cred(namespace=namespace)
+
+    # If `name` is a URL, assume it points to a cloud storage
+    storage_path = name if "://" in name else sp
+    storage_path = f"{storage_path.strip('/')}/{base_dir}"
+    tb_storage_uri = f"tiledb://{namespace}/{storage_path}"
+    logger.debug("Output storage path: %s", storage_path)
+    logger.debug("TileDB Storage URI: %r", tb_storage_uri)
+
+    access_credentials_name = access_credentials_name or acn
+    logger.debug("Storage path: %r", storage_path)
+    logger.debug("ACN: %s", access_credentials_name)
+
+    # Group check and/or creation
+    if "://" not in name:
+        group_uri = output_uri
+    else:
+        group_uri = f"tiledb://{namespace}/{base_dir}"
+    logger.debug("Group URI: %r", group_uri)
+
+    group_created = False
+    if not tiledb.object_type(group_uri, ctx=tiledb.cloud.Ctx()) == "group":
+        group_namespace, group_name = utils.split_uri(group_uri)
+        groups.create(
+            name=group_name,
+            namespace=group_namespace,
+            storage_uri=storage_path,
+            parent_uri=parent_group_uri,
+            credentials_name=access_credentials_name,
+        )
+        group_created = True
+        logger.debug("Group URI: %r created", group_uri)
+
+    # Upload Stats
+    logger.info("-----")
+    if parent_group_uri:
+        logger.info("Sub-Folder %r Upload Stats", base_dir)
+    else:
+        logger.info("Folder %r Upload Stats", base_dir)
+    logger.info("-----")
+    logger.info("- Input URI: %r", input_uri)
+    logger.info("- Output URI: %r", output_uri)
+    logger.info("-- Storage Path: %r", storage_path)
+    logger.info("- Group URI: %r", group_uri)
+    logger.info("-- Group Created: %s", group_created)
+
+    # Report results
+    uploaded = 0
+    dir_count = 0
+    upload_errors: Dict[str, str] = {}
+    results = UploadFoldersResults(
+        directory=base_dir, report="", errors={}, sub_folders=[]
+    )
+
+    vfs = tiledb.VFS(config=config)
+    # List local folder
+    input_ls: List[str] = vfs.ls(input_uri)
```

Comment:
There is also a `recursive=True` flag available, I think this might be needed to list the whole hierarchy below?

Reply:
@JohnMoutafis can you check the performance of the current implementation?

Reply:
Using the following snippet:

```python
def vfs_ls_recursive(uri):
    vfs = tiledb.VFS()
    files = []
    for fname in vfs.ls(uri):
        files.append(fname)
        if vfs.is_dir(fname):
            files += vfs_ls_recursive(fname)
    return files
```

to time the implementation's manual recursion against `vfs.ls(<folder URI>, recursive=True)` and `vfs.ls_recursive` on a local folder containing sub-folders, got the following results:
- `vfs_ls_recursive`: 1.18 ms ± 24.4 µs per loop (mean ± std. dev. of 100 runs, 1,000 loops each)
- `vfs.ls(<folder URI>, recursive=True)`: 1.95 ms ± 121 µs per loop (mean ± std. dev. of 100 runs, 1,000 loops each)
- `vfs.ls_recursive`: 1.96 ms ± 35.4 µs per loop (mean ± std. dev. of 100 runs, 1,000 loops each)

Timings were collected using [IPython's `%timeit`](https://ipython.readthedocs.io/en/stable/interactive/magics.html#magic-timeit) for 100 runs of 1000 loops per run.
The solutions seem comparable.